Support custom autograd functions in C++

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23572 Support custom autograd functions in C++**

Summary:
### **(The stack from #23020  was moved into this PR)**

Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. 

Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public Function<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->saved_data["n"] = n;
     return {var};
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->saved_data["n"].toInt();
     return {grad_output[0]*n};
  }
};

```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26). 

Test Plan:
Added tests for the custom autograd function API based on test_autograd.py. Currently only the tests for the basic functionality have been added. More tests will be added later.

Differential Revision: [D16583428](https://our.internmc.facebook.com/intern/diff/D16583428)

Commit:
Support custom autograd functions in C++

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Commit:
Update on "Support custom autograd functions in C++"


Summary:
### **(The stack from #23020  was moved into this PR)**

Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. 

Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public Function<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26). 

Test Plan:
Added tests for the custom autograd function API based on test_autograd.py. Currently only the tests for the basic functionality have been added. More tests will be added later.

Commit:
Update on "Support custom autograd functions in C++"


Summary:
### **(The stack from #23020  was moved into this PR)**

Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. 

Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public Function<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26). 

Test Plan:
Added tests for the custom autograd function API based on test_autograd.py. Currently only the tests for the basic functionality have been added. More tests will be added later.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -95,6 +95,8 @@ struct TORCH_API AutogradContext {
   variable_list to_save_;
 
+  // The CppNode in the autograd graph that owns this AutogradContext. We need a
+  // weak_ptr to avoid a refcycle.
```
I think this comment can be even better by explaining why `grad_fn_` will be live when you want to use it.

albanD commented on torch/csrc/api/include/torch/autograd.h:
```diff
@@ -0,0 +1,9 @@
+#include <torch/csrc/autograd/custom_function.h>
+
+namespace torch {
+  using Variable = autograd::Variable;
```
Why do these need to be exposed via torch:: now and not before? they look like autograd stuff no?

ezyang replied:
cc'ing @yf225, IIRC, the desired terminal end state is that we get rid of `Variable` entirely, is that right?

yf225 replied:
Yes that's correct. We should continue using `torch::autograd::Variable` before the terminal end state happens.

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +14,276 @@ TORCH_API variable_list _wrap_outputs(
   const at::ArrayRef<Variable> raw_outputs,
   const std::shared_ptr<Node> &cdata);
+
+// To use custom autograd operations implement a CFunction subclass with
```
nit: CFunction -> Function

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +14,276 @@ TORCH_API variable_list _wrap_outputs(
+//
+// backward() should take an AutogradContext* and a variable list containing as
+// many Variables as there were outputs from forward as arguments. It should
+//  return as many Variables as there were inputs with each of them containing
```
nit: extra space before "return"

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +14,276 @@ TORCH_API variable_list _wrap_outputs(
+//      var.mul_(2);
+//      // Mark var as modified by inplace operation
+//      ctx->mark_dirty({var});
+//      return std::vector<Variable>({var});
```
nit: std::vector<Variable> -> variable_list ?
And same for the backward if it is correct.

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +14,276 @@ TORCH_API variable_list _wrap_outputs(
+  bool requires_grad;
+};
+
+// Node representing the operation implemented by the user defined Function
```
This comment should be rephrased I think.

albanD commented on test/cpp/api/autograd.cpp:
```diff
@@ -0,0 +1,129 @@
+TEST(CustomAutogradTest, CustomFunction) {
+  struct MyFunction : public torch::autograd::Function<MyFunction> {
+    static variable_list forward(AutogradContext *ctx, Variable var1, int mul, Variable var2) {
+      ctx->saved_data["mul"] = mul;
```
Didn't that change to `save`?

malvika2147 replied:
It changed from `save` to `saved_data`. I'll update the comment.

ezyang replied:
One quick thought: I'm not looking forward to someone trying to save a Variable in `ctx->saved_data`. Does it error if you try to do this? Can we enforce that people not do this?

malvika2147 replied:
It does not enforce that only non variables are saved in saved_data, because the python API has the same behavior.

albanD reviewed (commented):
Looks mostly good !
Nice job.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -11,4 +14,276 @@ TORCH_API variable_list _wrap_outputs(
+  // The CppNode in the autograd graph that owns this AutogradContext. We need a
+  // weak_ptr to avoid a refcycle.
+  std::weak_ptr<Node> grad_fn_;
+  bool has_freed_buffers;
```
Trailing underscore please!

Commit:
Update on "Support custom autograd functions in C++"


Summary:
### **(The stack from #23020  was moved into this PR)**

Adding API for custom autograd operations, with user defined forward and backward, [like in python](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd).

The custom operation should be a subclass of Function, with static forward and backward functions. `forward()` can accept any arguments similar to the Python API and `backward()` should accept a variable list as an argument. 

Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public Function<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->saved_data["n"] = n;
     return {var};
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->saved_data["n"].toInt();
     return {grad_output[0]*n};
  }
};

```
Then, it can be used with:
```
Variable x;
MyFunction::apply(6, x);
```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26). 

Test Plan:
Added tests for the custom autograd function API based on test_autograd.py. Currently only the tests for the basic functionality have been added. More tests will be added later.
